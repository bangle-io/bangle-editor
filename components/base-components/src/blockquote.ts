import type { MarkdownSerializerState } from 'prosemirror-markdown';

import type { RawPlugins, RawSpecs } from '@bangle.dev/core';
import {
  Command,
  DOMOutputSpec,
  EditorState,
  keymap,
  Node,
  Schema,
  wrapIn,
  wrappingInputRule,
} from '@bangle.dev/pm';
import {
  copyEmptyCommand,
  cutEmptyCommand,
  moveNode,
} from '@bangle.dev/pm-commands';
import {
  createObject,
  filter,
  findParentNodeOfType,
  insertEmpty,
} from '@bangle.dev/utils';

import { getNodeType, getParaNodeType } from './helpers';

export const spec = specFactory;
export const plugins = pluginsFactory;
export const commands = {
  queryIsBlockquoteActive,
  wrapInBlockquote,
};
export const defaultKeys = {
  wrapIn: 'Ctrl-ArrowRight',
  moveDown: 'Alt-ArrowDown',
  moveUp: 'Alt-ArrowUp',
  emptyCopy: 'Mod-c',
  emptyCut: 'Mod-x',
  insertEmptyParaAbove: 'Mod-Shift-Enter',
  insertEmptyParaBelow: 'Mod-Enter',
};

const name = 'blockquote';

function specFactory(): RawSpecs {
  return {
    type: 'node',
    name,
    schema: {
      content: 'block*',
      group: 'block',
      defining: true,
      draggable: false,
      parseDOM: [{ tag: 'blockquote' }],
      toDOM: (): DOMOutputSpec => {
        return ['blockquote', 0];
      },
    },
    markdown: {
      toMarkdown: (state: MarkdownSerializerState, node: Node) => {
        state.wrapBlock('> ', null, node, () => state.renderContent(node));
      },
      parseMarkdown: {
        blockquote: {
          block: name,
        },
      },
    },
  };
}

function pluginsFactory({
  markdownShortcut = true,
  keybindings = defaultKeys,
} = {}): RawPlugins {
  return ({ schema }) => {
    const type = getNodeType(schema, name);
    return [
      markdownShortcut && wrappingInputRule(/^\s*>\s$/, type),
      keybindings &&
        keymap(
          createObject([
            [keybindings.wrapIn, wrapInBlockquote()],
            [keybindings.moveUp, moveNode(type, 'UP')],
            [keybindings.moveDown, moveNode(type, 'DOWN')],

            [keybindings.emptyCopy, copyEmptyCommand(type)],
            [keybindings.emptyCut, cutEmptyCommand(type)],

            [keybindings.insertEmptyParaAbove, insertEmptyParaAbove()],
            [keybindings.insertEmptyParaBelow, insertEmptyParaBelow()],
          ]),
        ),
    ];
  };
}

export function queryIsBlockquoteActive() {
  return (state: EditorState) => {
    const type = getNodeType(state, name);
    return Boolean(findParentNodeOfType(type)(state.selection));
  };
}

export function wrapInBlockquote() {
  return filter(
    (state) => !queryIsBlockquoteActive()(state),
    (state, dispatch, _view) => {
      const type = getNodeType(state, name);

      return wrapIn(type)(state, dispatch);
    },
  );
}

export function insertEmptyParaAbove(): Command {
  const isInBlockquote = queryIsBlockquoteActive();
  return (state, dispatch, view) => {
    const para = getParaNodeType(state);

    return filter(isInBlockquote, insertEmpty(para, 'above', true))(
      state,
      dispatch,
      view,
    );
  };
}

export function insertEmptyParaBelow(): Command {
  const isInBlockquote = queryIsBlockquoteActive();
  return (state, dispatch, view) => {
    const para = getParaNodeType(state);

    return filter(isInBlockquote, insertEmpty(para, 'below', true))(
      state,
      dispatch,
      view,
    );
  };
}
