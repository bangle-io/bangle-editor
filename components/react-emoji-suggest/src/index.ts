import * as emojiSuggest from './emoji-suggest';
export * from './EmojiSuggest';
export { emojiSuggest };

export type { GetEmojiGroupsType } from './emoji-suggest';
export { selectEmoji } from './emoji-suggest';
export { getSquareDimensions, resolveCounter, resolveRowJump } from './utils';
