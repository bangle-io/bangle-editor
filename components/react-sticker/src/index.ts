import * as sticker from './sticker';

export { sticker };
