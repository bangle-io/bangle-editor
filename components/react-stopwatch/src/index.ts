import * as stopwatch from './stopwatch';

export { stopwatch };
