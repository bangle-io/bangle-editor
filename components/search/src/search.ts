import type { RawPlugins } from '@bangle.dev/core';
import type { Command, EditorProps, Node } from '@bangle.dev/pm';
import {
  Decoration,
  DecorationSet,
  EditorState,
  Plugin,
  PluginKey,
} from '@bangle.dev/pm';
import { matchAllPlus } from '@bangle.dev/utils';

const name = 'search';
export const plugins = pluginsFactory;

function pluginsFactory({
  key = new PluginKey(name),
  query: initialQuery,
  className = 'bangle-search-match',
  maxHighlights = 1500,
}: {
  key: PluginKey<{
    query: string;
    decos: DecorationSet;
  }>;
  query?: RegExp | string;
  className?: string;
  maxHighlights?: number;
}): RawPlugins {
  function buildDeco(state: EditorState, query?: RegExp | string) {
    if (!query) {
      return DecorationSet.empty;
    }
    const matches = findMatches(state.doc, query, maxHighlights);
    const decorations = matches.map((match) => {
      // TODO we should improve the performance
      // by only creating decos which need an update
      // see https://discuss.prosemirror.net/t/how-to-update-multiple-inline-decorations-on-node-change/1493
      return Decoration.inline(
        match.pos + match.match.start,
        match.pos + match.match.end,
        {
          class: className,
        },
      );
    });

    return DecorationSet.create(state.doc, decorations);
  }
  return () =>
    new Plugin({
      key: key,
      state: {
        init(_, state) {
          return {
            query: initialQuery,
            decos: buildDeco(state, initialQuery),
          };
        },
        apply(tr, old, oldState, newState) {
          const meta = tr.getMeta(key);
          if (meta) {
            const newQuery = meta.query;
            return {
              query: newQuery,
              decos: buildDeco(newState, newQuery),
            };
          }
          return tr.docChanged
            ? {
                query: old.query,
                decos: buildDeco(newState, old.query),
              }
            : old;
        },
      },
      props: {
        decorations(state) {
          return key.getState(state)?.decos || null;
        },
      } as EditorProps,
    });
}

function findMatches(doc: Node, regex: RegExp | string, maxHighlights: number) {
  let results: {
    pos: number;
    match: ReturnType<typeof matchAllPlus>[0];
  }[] = [];
  let count = 0;
  let gRegex: RegExp;
  if (regex instanceof RegExp) {
    let flags = 'g';
    if (regex.ignoreCase) {
      flags += 'i';
    }
    if (regex.multiline) {
      flags += 'm';
    }
    gRegex = RegExp(regex.source, flags);
  } else {
    gRegex = RegExp(regex, 'g');
  }

  doc.descendants((node, pos) => {
    if (maxHighlights <= count) {
      return false;
    }
    if (node.isText) {
      const source = node.textContent;

      const matchedResult = matchAllPlus(gRegex, source);
      for (const match of matchedResult) {
        if (!match.match) {
          continue;
        }
        if (maxHighlights <= count++) {
          break;
        }
        results.push({
          pos,
          match,
        });
      }
    }
    return;
  });

  return results;
}

export function updateSearchQuery(
  key: PluginKey,
  query?: RegExp | string,
): Command {
  return (state, dispatch, _view) => {
    if (dispatch) {
      dispatch(state.tr.setMeta(key, { query }).setMeta('addToHistory', false));
    }
    return true;
  };
}
