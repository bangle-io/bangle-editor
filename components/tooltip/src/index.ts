import * as selectionTooltip from './selection-tooltip';
import type { SuggestTooltipRenderOpts } from './suggest-tooltip';
import * as suggestTooltip from './suggest-tooltip';
import type { TooltipRenderOpts } from './tooltip-placement';
import * as tooltipPlacement from './tooltip-placement';

export * from './create-tooltip-dom';
export { tooltipPlacement, selectionTooltip, suggestTooltip };
export type { TooltipRenderOpts };
export type { SuggestTooltipRenderOpts };
export type { SelectionTooltipProps } from './selection-tooltip';
