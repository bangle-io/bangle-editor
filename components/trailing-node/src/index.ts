import * as trailingNode from './trailing-node';

export { trailingNode };
