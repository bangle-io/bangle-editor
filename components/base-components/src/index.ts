// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item-component';
import * as orderedList from './ordered-list';
import * as image from './image';

export {
  doc,
  paragraph,
  text,
  history,
  editorStateCounter,
} from '@bangle.dev/core';

export {
  bold,
  code,
  italic,
  link,
  strike,
  underline,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  image,
};
