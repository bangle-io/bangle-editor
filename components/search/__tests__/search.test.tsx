/**
 * @jest-environment jsdom
 */
/** @jsx psx */

import { PluginKey, SpecRegistry } from '@bangle.dev/core';
import { defaultSpecs } from '@bangle.dev/all-base-components';
import { psx, renderTestEditor } from '@bangle.dev/test-helpers';
import { search } from '../src/index';

const specRegistry = new SpecRegistry([...defaultSpecs()]);

const searchPluginKey = new PluginKey();
function countOcurrences(string: string, match: string) {
  return string.split(match).length - 1;
}

test('shows decoration', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [search.plugins({ key: searchPluginKey, query: /hello/ })],
  });

  const { container, view } = testEditor(
    <doc>
      <para>hello [world]</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(1);

  expect(container).toMatchSnapshot();
});

test('1 shows multiple decorations', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [search.plugins({ key: searchPluginKey, query: /hello/ })],
  });

  const { container } = testEditor(
    <doc>
      <para>hello [world] hello world</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(2);

  expect(container).toMatchSnapshot();
});

test('2 shows multiple decorations', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [search.plugins({ key: searchPluginKey, query: 'o' })],
  });

  const { container } = testEditor(
    <doc>
      <para>hello [world] hello world</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(4);

  expect(container).toMatchSnapshot();
});

test('respects maxHighlights limit', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [
      search.plugins({ key: searchPluginKey, query: 'o', maxHighlights: 2 }),
    ],
  });

  const { container } = testEditor(
    <doc>
      <para>hello [world] hello world</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(2);

  expect(container).toMatchSnapshot();
});

test('respect case senstivity provided by regex', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [
      search.plugins({
        key: searchPluginKey,
        query: /hello/,
      }),
    ],
  });

  const { container, view } = testEditor(
    <doc>
      <para>Hello [world]</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(0);
});

test('ignores case when i flag is provided', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [
      search.plugins({
        key: searchPluginKey,
        query: /hello/i,
      }),
    ],
  });

  const { container, view } = testEditor(
    <doc>
      <para>Hello [world]</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(1);
});

test('1 works when no match', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [search.plugins({ key: searchPluginKey, query: /hello/ })],
  });

  const { container } = testEditor(
    <doc>
      <para>bye [world] bye world</para>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(0);
});

test('2 works when no match', async () => {
  const testEditor = renderTestEditor({
    specRegistry,
    plugins: [search.plugins({ key: searchPluginKey, query: /hello/ })],
  });

  const { container } = testEditor(
    <doc>
      <para>bye bye world</para>
      <heading>bye bye world</heading>
    </doc>,
  );

  expect(countOcurrences(container.innerHTML, '"bangle-search-match"')).toBe(0);
});
