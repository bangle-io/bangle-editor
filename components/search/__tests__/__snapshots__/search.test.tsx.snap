// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 shows multiple decorations 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    translate="no"
  >
    <p>
      <span
        class="bangle-search-match"
      >
        hello
      </span>
       world 
      <span
        class="bangle-search-match"
      >
        hello
      </span>
       world
    </p>
  </div>
</div>
`;

exports[`2 shows multiple decorations 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    translate="no"
  >
    <p>
      hell
      <span
        class="bangle-search-match"
      >
        o
      </span>
       w
      <span
        class="bangle-search-match"
      >
        o
      </span>
      rld hell
      <span
        class="bangle-search-match"
      >
        o
      </span>
       w
      <span
        class="bangle-search-match"
      >
        o
      </span>
      rld
    </p>
  </div>
</div>
`;

exports[`respects maxHighlights limit 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    translate="no"
  >
    <p>
      hell
      <span
        class="bangle-search-match"
      >
        o
      </span>
       w
      <span
        class="bangle-search-match"
      >
        o
      </span>
      rld hello world
    </p>
  </div>
</div>
`;

exports[`shows decoration 1`] = `
<div
  data-testid="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    translate="no"
  >
    <p>
      <span
        class="bangle-search-match"
      >
        hello
      </span>
       world
    </p>
  </div>
</div>
`;
